doc/design/custom_xcode_schemes.md

```diff
@@ -0,0 +1,467 @@
+
+The above declaration generates two schemes: `Foo` and
+`FooTests.__internal__.__test_bundle`. The `Foo` scheme contains configuration that builds
+`//Sources/Foo`, but has no configuration for test, launch or any of the other actions. The
+`FooTests.__internal__.__test_bundle` scheme contains configuration that builds `//Tests/FooTests`
+and executes `//Tests/FooTests` when testing is requested.
+
+## Introduction of Custom Xcode Schemes
+
+A new Starlark module called, `xcode_schemes`, provides functions for defining custom, Xcode
+schemes.
+
+### Defining a Simple Xcode Scheme
+
+Building on the previous example, let's define a scheme that combines the two targets,
+`//Sources/Foo` and `//Tests/FooTests`.
+
+```python
+load(
+    "@com_github_buildbuddy_io_rules_xcodeproj//xcodeproj:xcodeproj.bzl",
+    "xcode_schemes",
+    "xcodeproj",
+)
+
+_SCHEMES = [
+    xcode_schemes.scheme(
+        name = "Foo Module",
+        build_action = xcode_schemes.build_action(["//Sources/Foo"]),
+        test_action = xcode_schemes.test_action(["//Tests/FooTests"]),
+    ),
+]
+
+xcodeproj(
+    name = "generate_xcodeproj",
+    project_name = "Command Line",
+    tags = ["manual"],
+    schemes = _SCHEMES,
+)
+```
+
+The `xcode_schemes.scheme` function call defines a scheme with a user visible name of `Foo Module`.
+It is configured to build `//Sources/Foo` and `//Tests/FooTests`. (Targets listed in the
+test action and launch action are automatically added to the build action.) The scheme is also
+configured to execute the `//Tests/FooTests` target when testing is requested.
+
+The result of the function is wrapped in a `list` and passed to the `schemes` parameter of the
+`xcodeproj` macro.
+
+### Specifying a Launch Target
+
+Let's continue our example. We will add an `ios_application` and a `ios_ui_test` to the mix.
+
+```python
+# Assumptions
+#   //Sources/Foo:Foo - swift_library
+#   //Sources/FooTests:FooTests = ios_unit_test
+#   //Sources/App = ios_application
+#   //Sources/AppUITests = ios_ui_test
+
+load(
+    "@com_github_buildbuddy_io_rules_xcodeproj//xcodeproj:xcodeproj.bzl",
+    "xcode_schemes",
+    "xcodeproj",
+)
+
+_SCHEMES = [
+    xcode_schemes.scheme(
+        name = "Foo Module",
+        build_action = xcode_schemes.build_action(["//Sources/Foo"]),
+        test_action = xcode_schemes.test_action(["//Tests/FooTests"]),
+    ),
+    xcode_schemes.scheme(
+        name = "My Application",
+        test_action = xcode_schemes.test_action([
+            "//Tests/AppUITests",
+            "//Tests/FooTests",
+        ]),
+        launch_action = xcode_schemes.launch_action("//Sources/App"),
+    ),
+]
+
+xcodeproj(
+    name = "generate_xcodeproj",
+    project_name = "Command Line",
+    tags = ["manual"],
+    schemes = _SCHEMES,
+)
+```
+
+The above example adds a second scheme called `My Application`. It is configured with two test targets
+and a single launch target. All three targets are implicitly included in the build list.
```

Comment:
>All three targets are implicitly included in the build list.

Is `//Tests/AppUITests` set to build for `Run` or only `Test` (default behavior would only be `Test`). And are we able to force it to build during `Run` by listing it explicitly in `build_action`?

Reply:
Good question. For reference, here is the [BuildFor enum](https://github.com/tuist/XcodeProj/blob/3a93b47a34860a4d7dbcd9cc0ae8e9543c179c61/Sources/XcodeProj/Scheme/XCScheme%2BBuildAction.swift#L8-L13).

I was thinking that we would just set the `buildFor` to `default`.  For autogenerated schemes, we set the following list:
```swift
                buildFor: [
                    .running,
                    .testing,
                    .profiling,
                    .archiving,
                    .analyzing
                ]
```

I think that we can be as smart of dumb about this as we want. If you think that it is important to be selective, we can add it. I don't have a good sense for whether folks customize this for their real projects.

Reply:
I believe Xcode defaults to having test targets `buildFor: [.testing]`.

I say anything we do this:
- Anything `build_action` gets `.testing` in `buildFor`
- Anything in `launch_action` or `build_action` gets `[.running, .profiling, .archiving, .analyzing]` in `buildFor`
- Thus being in both gets all of them

Reply:
Did you mean `test_action` for the first bullet?

Reply:
The rest sounds reasonable to me.

Reply:
I added a section enumerating the build for logic.